src/routes/guides/testing-your-app.mdx

```diff
@@ -0,0 +1,404 @@
+  ...renderResults, // see render results
+} = renderResult
+```
+
+The `container` from the `...renderResults` will contain the targetElement (default is a `<div>`). That and the ability to manipulate the `arg` signal are helpful in testing directives.
+
+If for example there is a directive that handles the Fullscreen API, it could be tested like this:
+
+<TabsCodeBlocks>
+  <div id="fullscreen.test.ts">
+```ts frame="none"
+import { test, expect, vi } from "vitest"
+import { renderDirective } from "@solidjs/testing-library"
+import { createFullScreen } from "./fullscreen"
+
+test("toggles fullscreen", () => {
+  const targetElement = document.createElement("div")
+  const fs = vi.spyOn(targetElement, "fullscreen")
+  const [setArg, container] = renderDirective(createFullScreen, false)
+  setArg(true)
+  expect(fs).toHaveBeenCalled()
+})
+```
+  </div>
+  <div id="fullscreen.ts">
+```ts frame="none"
+import { Accessor } from "solid-js"
+
+export const fullscreen = (ref: HTMLElement, active: Accessor<boolean>) =>
+  createEffect(() => {
+    const isActive = document.fullscreenElement === ref
+    if (active() && !isActive) {
+      ref.requestFullScreen().catch(() => {})
+    } else if (!active() && isActive) {
+      document.exitFullScreen()
+    }
+  })
+```
+  </div>
+</TabsCodeBlocks>
+
+## Primitive testing
+
+If the reference to the element is not needed, parts of state and logic can be put into reusable hooks or primitives. Since those do not require any elements, using `render` to test them would require a component that otherwise has no use. To circumvent that, there is a `renderHook` utility that simulates such a component without actually rendering anything.
+
+```ts frame="none"
+const renderResult = renderHook(hook, {
+  initialProps, // an array with arguments being supplied to the hook
+  wrapper, // same as the wrapper optionss for `render`
+})
+const {
+  result, // return value of the hook (mutable, destructuring fixes it)
+  cleanup, // manually remove the traces of the test from the DOM
+  owner, // the owner running the hook to use with `runWithOwner()` 
+} = renderResult
+```
+
+A primitive that manages the state of a counter could be tested like this:
+
+```ts frame="none"
+import { test, expect } from "vitest"
+import { renderHook } from "@solidjs/testing-library"
+import { createCounter } from "./counter"
+
+test("increments count", () => {
+  const { result } = renderHook(createCounter)
+  expect(result.count).toBe(0)
+  result.increment()
+  expect(result.count).toBe(1)
+})
+```
+
+
+## Testing effects
+
+Since effects may happen asynchronous, it can be difficult to test them. `@solidjs/testing-library` comes with a `testEffect` function that takes another function, which will receive a `done` function that can be called once the tests are over. It returns a promise that is resolved once done is called. Errors that would hit the next boundary are used to reject the returned promise.
```

Comment:
```suggestion
Since effects may happen asynchronously, it can be difficult to test them. 
[`@solidjs/testing-library` comes with a `testEffect` function](https://testing-library.com/docs/solid-testing-library/api#async-methods) that takes another function that receives a `done` function to be called once tests are over.
Once `done` is called, it returns a promise that is resolved.
Any errors that would hit the next boundary are used to reject the returned promise.
```

Reply:
It is not done that returns the promise, but testEffect; done only triggers its resolution.